### PR TITLE
Feature #90: adding TLS mutual authentication to vertica-sql-go

### PR DESCRIPTION
**Pull request by bmyers-csu**
Feature #90 
Add support for custom TLS configurations implemented in the same structure as mysql Go client.
@mjskier contributed as well

Some code taken from @ehrt74 and this PR will replace #91

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -569,21 +605,23 @@ func (v *connection) initializeSSL(sslFlag string) error {
...
 		connectionLogger.Info("enabling SSL/TLS server strict mode")
 		v.conn = tls.Client(v.conn, &tls.Config{ServerName: v.connURL.Hostname()})
 	default:
-		err := fmt.Errorf("unsupported tlsmode flag: %s - should be 'server', 'server-strict' or 'none'", sslFlag)
-		connectionLogger.Error(err.Error())
-		return err
+		connectionLogger.Info("enabling SSL/TLS custom mode")
+		config, ok := tlsConfigs.get(sslFlag)
+		if !ok {
+			err := fmt.Errorf("tls config %s not registered", sslFlag)
```
Comment (sitingren) [inline connection.go:618]: I think it is better to provide guidance about how to register with "RegisterTLSConfig()" in this error message.

File: connection.go [hunk 2]
```diff
@@ -569,21 +605,23 @@ func (v *connection) initializeSSL(sslFlag string) error {
...
 		v.conn = tls.Client(v.conn, &tls.Config{InsecureSkipVerify: true})
-	case "server-strict":
+	case tlsModeServerStrict:
 		connectionLogger.Info("enabling SSL/TLS server strict mode")
 		v.conn = tls.Client(v.conn, &tls.Config{ServerName: v.connURL.Hostname()})
 	default:
-		err := fmt.Errorf("unsupported tlsmode flag: %s - should be 'server', 'server-strict' or 'none'", sslFlag)
-		connectionLogger.Error(err.Error())
-		return err
+		connectionLogger.Info("enabling SSL/TLS custom mode")
```
Comment (sitingren) [inline connection.go:615]: Adding a comment to indicate this is the implementation for "mutual ssl mode" will be helpful for other developers in the future.

File: resources/tests/genCerts.sh [hunk 3]
```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
```
Comment (sitingren) [inline resources/tests/genCerts.sh:1]: Would you tell me how stable it is compared to using our current hard-coded ssl files? Our CI tests (.github/workflows/ci.yaml) rely on the `ubuntu-latest` image. The version of commands (e.g. openssl) on the image will not be changeless in the future. Is there a risk that this file will break?

File: driver_test.go [hunk 4]
```diff
@@ -1022,26 +1044,26 @@ func TestEnableResultCache(t *testing.T) {
 	testEnableResultCachePageSized(t, connDB, vCtx, 0)
 }
 
-//func TestConnectionClosure(t *testing.T) {
-//	adminDB := openConnection(t, "test_connection_closed_pre")
-//	defer closeConnection(t, adminDB, "test_connection_closed_post")
-//	const userQuery = "select 1 as test"
+// func TestConnectionClosure(t *testing.T) {
```
Comment (sitingren) [inline driver_test.go:1047]: Need to undo extra whitespaces for this function.

File: connection.go [hunk 5]
```diff
@@ -56,6 +56,42 @@ var (
 	connectionLogger = logger.New("connection")
 )
 
+const (
```
Comment (sitingren) [inline connection.go:59]: I like the idea to make them constants!

**Review by sitingren — COMMENTED** [review]
Thanks for the effort to put them together! It looks good overall. See comments for my concern about testing.

File: resources/tests/genCerts.sh (hunk 3 above)
Comment (mjskier) [reply resources/tests/genCerts.sh:1]: My train of thought on this:

Hard coded certs are black boxes with no hint as to how they were generated. And if they need to change that makes it harder for somebody to create new certs by hand. (for example for my own system I found out that for Vertica 10.x I needed to add `digitalSignature` to keyUsage for the certs to work. That was not required with vertica 9.x. It was trivial for me to add that to my script, as opposed to chasing the right combination of commands to create certs.

I have to think that the `openssl` API is pretty stable, and any change to the existing options would meet with resistance considering how many products document the sequence of commands used in this script.

Comment (sitingren) [reply resources/tests/genCerts.sh:1]: @mjskier Make sense. Thanks for the explanation!

File: resources/tests/genCerts.sh [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+: ${COUNTRY:=US}
+: ${STATE:=CO}
+: ${LOCALE:=FortCollins}
+: ${COMPANY:=Vertica}
+: ${ORG:=SQL}
+: ${CN:=localhost}
+
+: ${CERT_LOC:=./resources/tests/ssl}￼
```
Comment (sitingren) [inline resources/tests/genCerts.sh:10]: What is this ￼?

Comment (mjskier) [reply resources/tests/genCerts.sh:10]: I'm sorry, I don't understand the question. Github shows me a box with OBJ in it before the question mark.
Are you asking about the value of these fields, or the construct to get their values, or...

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -27,21 +28,37 @@ jobs:
...
+
+      - name: Generating certs
+        run: |
+          ./resources/tests/genCerts.sh
+
       - name: Run tests
         run: |
           go test -race . -args --locator localhost:5433 --user dbadmin
-          go test -race . -args --locator localhost:5433 --user dbadmin --use_prepared_statements=0
+          # go test -race . -args --locator localhost:5433 --user dbadmin --use_prepared_statements=0
```
[the remark this answers is not in the document]
Comment (mjskier) [reply .github/workflows/ci.yaml]: Done

File: driver_test.go (hunk 4 above)
Comment (mjskier) [reply driver_test.go]: Done

File: connection.go (hunk 1 above)
Comment (mjskier) [reply connection.go]: done (referred to the README.md which has an example on how to register and use a custom config)

File: connection.go (hunk 2 above)
Comment (mjskier) [reply connection.go:616]: Done. Good suggestion, I agree that "custom TLS" doesn't have any hint that it is used for mutual ssl mode.

File: resources/tests/genCerts.sh (hunk 6 above)
Comment (sitingren) [reply resources/tests/genCerts.sh:10]: Github shows me a box with OBJ in it. Is this a meaningful value or a typo?

Comment (sitingren) [reply resources/tests/genCerts.sh]: That's the only place I saw this symbol. It is visible in Sublime with UTF-8 encoding from my end. Probably this is a matter of the setting of your editor.

File: README.md [hunk 8]
```diff
@@ -112,6 +113,40 @@ err = connDB.PingContext(ctx)
...
+pem, err := ioutil.ReadFile("/certs/ca.crt")
+if err != nil {
+    LOG.Warningln("ERROR: failed reading cert file", err)
+}
+if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
+    LOG.Warningln("ERROR: Failed to append PEM")
+}
+tlsConfig := &tls.Config{RootCAs: rootCertPool, ServerName: host}
+vertigo.RegisterTLSConfig("myCustomName", tlsConfig)
+}
```
Comment (sitingren) [inline README.md:135]: Syntax error: extra `}`

Comment (mjskier) [reply README.md]: done

**Comment by sitingren** [comment]
Merged. Thanks for the contribution!

**Pull request merged**